goal_src/goal-lib.gc

```diff
@@ -756,6 +768,34 @@
   `(new 'debug 'pair ,a ,b)
   )
 
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; ARRAYS
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defmacro first-arr (coll)
+  "Returns the first element in an array"
+  `(-> ,coll 0))
+
+(defmacro last-arr (coll)
+  "Returns the last element in an array"
+  `(-> ,coll (dec (length ,coll))))
+
+(defmacro last-idx-arr (coll)
+  "Returns the index of the last element in an array"
+  `(dec (length ,coll)))
+
+(defmacro arr-idx-of (coll val def)
+  "Returns the index of an item in an array, returns <def> if is nothing is found.  Assumes the collection is non-empty"
+  `(let ((curr-val #f)
+         (curr-idx -1))
+    (until (or (> curr-idx (length ,coll))
+               (= curr-val ,val))
+      (set! curr-idx (inc curr-idx))
+      (set! curr-val (-> ,coll curr-idx)))
```

Comment:
I think this can read 2 past the end - the `or` check will let an iteration run with `curr-idx` equal to the length, then it increments past that.  This is super awkward to write without an early return, but OpenGOAL does support named blocks to return from, like this:
```
(block find-element
  (dotimes (i (lenth arr))
     (if (= val (-> arr i))
      (return-from find-element i)))
  def)
```
you could be the first to use this feature!

Reply:
ah i originally wanted to do an early-return version but i couldn't figure it out. This looks much better.